Module4_Concepts/BinarySearchTreeDelete.md

```diff
@@ -0,0 +1,59 @@
+
+#### BSTDelete()
+
+Due to the various cases when deleting a BST node, the `BSTDelete` function is a little more complex than `BSTSearch` and `BSTInsert`. However, don't be intimidated by the code; make certain you understand the process of deleting a node, because the code simply follows that logic. Let's take a look:
+
+```Python
+def smallestNode(curNode):
+    inspectedNode = curNode
+    while(inspectedNode.left is not None):
+        inspectedNode = inspectedNode.left  #Go left as far as possible
+    return inspectedNode
+
+def BSTDelete(curNode, key):
+    if curNode is None:
+        return curNode
+    if (key < curNode.key):
+        curNode.left = BSTDelete(curNode.left, key)
+    elif (key > curNode.key):
+        curNode.right = BSTDelete(curNode.right, key)
+    else:  #found node to be deleted
+        if curNode.left is None:  #Case 1
+            tempNode = curNode.right
+            curNode = None  #Delete node
+            return tempNode
+        elif curNode.right is None:  #Case 1
+            tempNode = curNode.left
+            curNode = None  #Delete node
+            return tempNode
+        	
+        #Internal Node Case:
+        tempNode = smallestNode(curNode.right) #find smallest key node of right subtree
+        curNode.key = tempNode.key
+        curNode.right = BSTDelete(curNode.right, tempNode.key)
+        
+            
+```
+
+The first half of `BSTDelete` is simply finding the node that is desired to be deleted. Once found, it will enter the else statement and begin the process of being deleted. The code becomes a bit confusing when we get to the **Internal Node Case**. As you can see, it takes advantage of a function `smallestNode` that simply goes left until it finds null, which will return the smallest key in that tree. The **Internal Node Case** calls smallestNode on the right child of the node we want to delete, meaning we are going to be replacing the node we want to delete with the smallest key node of its right subtree. We then simply swap their keys and call `BSTDelete` on the node now containing the key of the node we orginally wanted to delete (this node will eventually be deleted once it gets to a **Case 1** scenario). 
+
+Although `BSTDelete` seems complicated initially, if you read through it slowly, you will realize that it simply follows the logic of either the **Leaf/1 Child Case** or the **Internal Node Case**. I recommend rereading the two cases if you're having trouble understanding the `BSTDelete` code.
+
+The Big O time complexity of `delete()` is O(h), h being the height of the binary search tree. So, the 'taller' the tree, the more time it takes to delete a value. The node that takes the longest to delete would the deepest leaf node, or the one farthest away from the root.
```

Comment:
the node that takes the longest to delete would also be the root, because it takes O(h) to replace the root with the smallest. you have to take into account both the deletion action itself and the erplacement.

also what is h in terms of n?